dask_ml/model_selection/_incremental.py

```diff
@@ -347,3 +394,548 @@ def fit(
+    error_score : "raise" # TODO: remove
+    return_train_socre : bool # TODO: remove
+    scheduler : # TODO: remove
+    n_jobs : int # TODO: remove
+    cache_cv : bool : # TODO: remove
+
+    Examples
+    --------
+    Connect to the client and create the data
+
+    >>> from dask.distributed import Client
+    >>> client = Client()
+    >>> import numpy as np
+    >>> from dask_ml.datasets import make_classification
+    >>> X, y = make_classification(n_samples=5000000, n_features=20,
+    ...                            chunks=100000, random_state=0)
+
+    Our underlying estimator is an SGDClassifier. We specify a few parameters
+    common to each clone of the estimator.
+
+    >>> from sklearn.linear_model import SGDClassifier
+    >>> model = SGDClassifier(tol=1e-3, penalty='elasticnet', random_state=0)
+
+    The distribution of parameters we'll sample from.
+
+    >>> params = {'alpha': np.logspace(-2, 1, num=1000),
+    ...           'l1_ratio': np.linspace(0, 1, num=1000),
+    ...           'average': [True, False]}
+
+    >>> search = RandomizedIncrementalSearch(model, params, random_state=0)
+    >>> search.fit(X, y, classes=[0, 1])
+    RandomizedIncrementalSearch(...)
+    """
+
+    def __init__(
+        self,
+        estimator,
+        param_distribution,
+        n_iter=10,
+        max_iter=100,
+        patience=10,
+        tol=0.001,
+        test_size=0.15,
+        random_state=None,
+        scoring=None,
+        iid=True,
+        refit=True,
+        error_score="raise",
+        return_train_score=_RETURN_TRAIN_SCORE_DEFAULT,
+        scheduler=None,
+        n_jobs=-1,
+        cache_cv=True,
+    ):
+        self.max_iter = max_iter
+        self.n_iter = n_iter
+        self.patience = patience
+        self.tol = tol
+        super(RandomizedIncrementalSearch, self).__init__(
+            estimator,
+            param_distribution,
+            test_size,
+            random_state,
+            scoring,
+            iid,
+            refit,
+            error_score,
+            return_train_score,
+            scheduler,
+            n_jobs,
+            cache_cv,
+        )
+
+    def _get_params(self):
+        return ParameterSampler(self.parameters, self.n_iter)
+
+    def _additional_calls(self, info):
+        out = {}
+        max_iter = self.max_iter
+        patience = self.patience
+        tol = self.tol
+
+        for ident, records in info.items():
+            if max_iter is not None and len(records) > max_iter:
+                out[ident] = 0
+
+            elif len(records) > patience:
+                old = records[-patience]["score"]
+                if all(d["score"] < old + tol for d in records[-patience:]):
+                    out[ident] = 0
+                else:
+                    out[ident] = 1
+
+            else:
+                out[ident] = 1
+        return out
+
+
+class SuccessiveReductionSearch(BaseIncrementalSearch):
```

Comment:
Could we make this class private? It's required for #221 but I'm not inclined to expose it.

Reply:
We're in a private submodule, and it's not imported in `model_selection/__init__.py`, so at the moment it's private.

Reply:
> I'm not inclined to expose it

Why not?

Reply:
Almost all of this class can be revealed through Hyperband if some "skip these brackets" interface is provided. This would allow the user to say "focus on the most exploratory brackets" and let Hyperband select the parameters.

Additionally, this class

* is implemented to allow Hyperband to work, and wouldn't be present otherwise
* has 3 hyper-params, and it's not clear how to trade off between them (`n_iter`, `max_iter` and `decay_rate`).

Reply:
> is implemented to allow Hyperband to work, and wouldn't be present otherwise

That's not necessarily true.  I like this class because it achieves the general objective of "train a lot at the beginning but focus on well-performing models" while also being simple and easy to understand.  I think that we should point people towards Hyperband in docs, but I don't see a reason to hide this implementation.

Reply:
I think that you're generally saying "but Hyperband performs better" and I'm saying "that's not the only thing being considered".

Reply:
Agreed. I think we should provide the parameters of this class in `HyperbandCV.metadata`, which will allow the user to use this class if they desire (and then we won't have to implement "skip this bracket").